gkctl/main.c

```diff
@@ -0,0 +1,218 @@
+
+static char doc[] = "Gatekeeper Client -- configure the Gatekeeper via "
+	"dynamic configuration";
+
+static struct argp_option options[] = {
+	{"server-path",		's',	"FILE",		0,
+		"Path to Gatekeeper/Grantor's UNIX socket",	1},
+	{ 0 }
+};
+
+struct args {
+	const char *server_path;
+	const char *lua_script_path;
+};
+
+static error_t
+parse_opt(int key, char *arg, struct argp_state *state)
+{
+	struct args *args = state->input;
+
+	switch (key) {
+	case 's':
+		args->server_path = arg;
+		break;
+
+	case ARGP_KEY_INIT:
+		args->lua_script_path = NULL;
+		break;
+
+	case ARGP_KEY_ARG:
+		if (args->lua_script_path) {
+			argp_error(state,
+				"Wrong number of arguments; only one is allowed");
+		}
+		args->lua_script_path = arg;
+		break;
+
+	case ARGP_KEY_END:
+		if (!args->lua_script_path) {
+			argp_error(state,
+				"The lua script path was not specified");
+		}
+		break;
+
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+
+	return 0;
+}
+
+static struct argp argp = {options, parse_opt, adoc, doc, NULL, NULL, NULL};
+
+static int
+load_file_to_buffer(const char *file_name, char *buffer, int n)
+{
+	int ret = 0;
+	FILE *fp;
+
+	memset(buffer, 0, n);
+
+	if ((fp = fopen(file_name, "r")) == NULL) {
+		fprintf(stderr, "Error: Failed to open file %s - %s\n",
+			file_name, strerror(errno));
+		ret = -1;
+		goto out;
+	}
+
+	/*
+	 * Return value equals the number of bytes transferred
+	 * only when size (i.e., second parameter) is 1.
+	 */
+	ret = fread(buffer, 1, n, fp);
+	if ((ret < n) && ferror(fp)) {
+		fprintf(stderr, "Error: Failed to read file %s - %s\n",
+			file_name, strerror(errno));
+		ret = -1;
+	}
+
+	fclose(fp);
+out:
+	return ret;
+}
+
+int
+main(int argc, char *argv[])
+{
+	int ret = 0;
+	int sock_fd;
+	char *line = NULL;
+	char file_str[MSG_MAX_LEN];
+	char send_buff[MSG_MAX_LEN + 2];
+	char recv_buff[MSG_MAX_LEN + 2];
+	size_t len = 0;
+	size_t total_file_len = 0;
+	uint16_t msg_len;
+	int send_size = 0;
+	int recv_size = 0;
+	int tot_send = 0;
+	int tot_recv;
+	struct sockaddr_un serv_addr; 
+
+    	struct args args = {
+		/* Defaults. */
+		.server_path = "/tmp/dyn_cfg.socket",
+	};
+
+	/* Read parameters. */
+	argp_parse(&argp, argc, argv, 0, NULL, &args);
+
+	if (sizeof(serv_addr.sun_path) < strlen(args.server_path)) {
+		fprintf(stderr, "Error: passing a too long server path - %s!\n",
+			args.server_path);
+		ret = -1;
+		goto out;
+	}
+
+	ret = load_file_to_buffer(args.lua_script_path,
+		file_str, sizeof(file_str));
+	if (ret <= 0) {
+		ret = -1;
+		goto out;
+	}
+	total_file_len = ret;
+	msg_len = htons(ret);
+
+	memcpy(send_buff, &msg_len, sizeof(msg_len));
+	memcpy(send_buff + sizeof(msg_len), file_str, total_file_len);
+
+	if((sock_fd = socket(AF_UNIX, SOCK_STREAM, 0)) < 0) {
+		perror("Error : Could not create socket\n");
+		ret = -1;
+		goto out;
+	}
+
+	memset(&serv_addr, 0, sizeof(serv_addr));
+	serv_addr.sun_family = AF_UNIX;
+	strncpy(serv_addr.sun_path, args.server_path,
+		strlen(args.server_path));
```

Comment:
We don't need the `memset()` here, just ensure that the string is terminated with `'\0'`.

Reply:
Move this part of the code to be just after the test in which we guarantee that `args.server_path` fits in `serv_addr.sun_path`, so it's obvious that we see the protection.